File: SyntheticData+AI/frontend/src/lib/offline/autoencoder.ts
/**
 * The autoencoder, evaluated in the browser from the exported JSON bundle.
 *
 * Same weights, same arithmetic, same threshold as the server. This is what
 * lets the published site work with no backend running at all -- and it is
 * only possible because training exports plain JSON instead of a pickle.
 */

export interface LayerJson {
  weight: number[][];
  bias: number[];
  activation: "relu" | "linear";
}

export interface BundleJson {
  schema_version: number;
  feature_names: string[];
  scaler: { mean: number[]; scale: number[] };
  layers: LayerJson[];
  residual_scale: number[] | null;
  threshold: number;
  metadata: Record<string, unknown>;
}

export const SUPPORTED_SCHEMA = 1;

interface Layer {
  weight: Float64Array[];
  bias: Float64Array;
  relu: boolean;
}

export class Autoencoder {
  private readonly layers: Layer[];
  readonly featureNames: string[];
  readonly mean: Float64Array;
  readonly scale: Float64Array;
  readonly residualScale: Float64Array | null;
  readonly threshold: number;
  readonly metadata: Record<string, unknown>;

  constructor(bundle: BundleJson) {
    if (bundle.schema_version !== SUPPORTED_SCHEMA) {
      throw new Error(
        `model bundle schema ${bundle.schema_version} is not supported (this build reads ${SUPPORTED_SCHEMA})`,
      );
    }
    this.featureNames = bundle.feature_names;
    this.mean = Float64Array.from(bundle.scaler.mean);
    this.scale = Float64Array.from(bundle.scaler.scale);
    this.residualScale = bundle.residual_scale
      ? Float64Array.from(bundle.residual_scale)
      : null;
    this.threshold = bundle.threshold;
    this.metadata = bundle.metadata ?? {};
    this.layers = bundle.layers.map((layer, index) => ({
      weight: layer.weight.map((row) => Float64Array.from(row)),
      bias: Float64Array.from(layer.bias),
      relu: layer.activation === "relu" && index < bundle.layers.length - 1,
    }));
  }

  get architecture(): number[] {
    return [this.featureNames.length, ...this.layers.map((l) => l.bias.length)];
  }

  get latentDim(): number {
    return Math.min(...this.layers.map((l) => l.bias.length));
  }

  get parameterCount(): number {
    return this.layers.reduce(
      (total, l) => total + l.bias.length * l.weight[0].length + l.bias.length,
      0,
    );
  }

  standardise(features: Float64Array): Float64Array {
    const out = new Float64Array(features.length);
    for (let i = 0; i < features.length; i++) {
      out[i] = (features[i] - this.mean[i]) / this.scale[i];
    }
    return out;
  }

  private forward(input: Float64Array): Float64Array {
    let current = input;
    for (const layer of this.layers) {
      const next = new Float64Array(layer.bias.length);
      for (let o = 0; o < layer.bias.length; o++) {
        const row = layer.weight[o];
        let sum = layer.bias[o];
        for (let i = 0; i < row.length; i++) sum += row[i] * current[i];
        next[o] = layer.relu ? Math.max(sum, 0) : sum;
      }
      current = next;
    }
    return current;
  }

  /** Per-feature squared reconstruction error, residual-normalised. */
  errors(features: Float64Array): Float64Array {
    const scaled = this.standardise(features);
    const reconstructed = this.forward(scaled);
    const out = new Float64Array(scaled.length);
    for (let i = 0; i < scaled.length; i++) {
      let residual = reconstructed[i] - scaled[i];
      if (this.residualScale) residual /= this.residualScale[i];
      out[i] = residual * residual;
    }
    return out;
  }

  score(features: Float64Array): number {
    const errors = this.errors(features);
    let total = 0;
    for (let i = 0; i < errors.length; i++) total += errors[i];
    return total / errors.length;
  }

  /** 0-100, hitting exactly 50 at the alarm threshold. Mirrors `health_index`. */
  healthIndex(score: number): number {
    if (this.threshold <= 0) return 0;
    const ratio = Math.max(score, 1e-12) / this.threshold;
    return Math.min(100, Math.max(0, 100 * 2 ** -(ratio ** 1.6)));
  }

  /** Which features the model failed hardest to reconstruct, and by how much. */
  topContributors(features: Float64Array, k = 4): Array<{ feature: string; share: number }> {
    const errors = this.errors(features);
    let total = 0;
    for (let i = 0; i < errors.length; i++) total += errors[i];
    total = total || 1;
    return Array.from(errors)
      .map((value, i) => ({ feature: this.featureNames[i], share: value / total }))
      .sort((a, b) => b.share - a.share)
      .slice(0, k);
  }

  /** Standardised deviation per feature -- "how many healthy sigmas out". */
  zScores(features: Float64Array): Record<string, number> {
    const out: Record<string, number> = {};
    this.featureNames.forEach((name, i) => {
      out[name] = (features[i] - this.mean[i]) / this.scale[i];
    });
    return out;
  }
}

File: SyntheticData+AI/src/vfactory/autoencoder.py
"""Portable autoencoder: a JSON weight bundle plus a NumPy forward pass.

Training happens in PyTorch (see :mod:`vfactory.train`), but nothing at serving
time needs an autograd engine. Exporting the fitted weights to plain JSON and
running the forward pass in NumPy means:

* the deployed API installs ``numpy`` instead of a ~900 MB deep-learning stack,
  so a free-tier container cold-starts in seconds rather than minutes;
* the exact same weights can be shipped to the browser and evaluated in
  TypeScript, which is how the static demo works with no backend at all;
* the artifact is a readable, diffable, version-controllable text file rather
  than a pickle -- ``torch.load`` on an untrusted ``.pth`` executes arbitrary
  code, which the original project did without ``weights_only``.

A round-trip test asserts the NumPy and PyTorch outputs agree, so the two
implementations cannot silently drift apart.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Literal

import numpy as np

SCHEMA_VERSION = 1

Activation = Literal["relu", "linear"]


def _apply(x: np.ndarray, activation: Activation) -> np.ndarray:
    if activation == "relu":
        return np.maximum(x, 0.0)
    if activation == "linear":
        return x
    raise ValueError(f"unsupported activation {activation!r}")


@dataclass(frozen=True)
class Layer:
    """A dense layer stored row-major as ``(out_features, in_features)``."""

    weight: np.ndarray
    bias: np.ndarray
    activation: Activation

    def forward(self, x: np.ndarray) -> np.ndarray:
        return _apply(x @ self.weight.T + self.bias, self.activation)

    def to_json(self) -> dict[str, Any]:
        return {
            "weight": self.weight.tolist(),
            "bias": self.bias.tolist(),
            "activation": self.activation,
        }

    @classmethod
    def from_json(cls, payload: dict[str, Any]) -> Layer:
        return cls(
            weight=np.asarray(payload["weight"], dtype=np.float64),
            bias=np.asarray(payload["bias"], dtype=np.float64),
            activation=payload["activation"],
        )


@dataclass(frozen=True)
class AutoencoderBundle:
    """Everything needed to score a feature vector: scaler, weights, threshold."""

    feature_names: tuple[str, ...]
    mean: np.ndarray
    scale: np.ndarray
    layers: tuple[Layer, ...]
    threshold: float
    metadata: dict[str, Any]
    #: Per-feature RMS reconstruction residual on healthy training data.
    #: Dividing by it before squaring turns the score into a Mahalanobis-like
    #: distance: features the network already reconstructs well (the structured
    #: ones) dominate the score, while features that are irreducible noise even
    #: on a healthy machine stop diluting it. Without this the mean-squared
    #: error over 26 features buries a defect signature under 25 noisy ones.
    residual_scale: np.ndarray | None = None

    # ---------------------------------------------------------------- #
    # Scoring
    # ---------------------------------------------------------------- #

    @property
    def n_features(self) -> int:
        return len(self.feature_names)

    @property
    def latent_dim(self) -> int:
        return int(min(layer.bias.shape[0] for layer in self.layers))

    @property
    def architecture(self) -> list[int]:
        return [self.n_features] + [int(layer.bias.shape[0]) for layer in self.layers]

    def standardise(self, features: np.ndarray) -> np.ndarray:
        return (np.atleast_2d(features) - self.mean) / self.scale

    def reconstruct(self, scaled: np.ndarray) -> np.ndarray:
        out = np.atleast_2d(scaled)
        for layer in self.layers:
            out = layer.forward(out)
        return out

    def residuals(self, features: np.ndarray) -> np.ndarray:
        """Signed reconstruction residual per feature, in standardised units."""
        scaled = self.standardise(features)
        return self.reconstruct(scaled) - scaled

    def errors(self, features: np.ndarray) -> np.ndarray:
        """Per-feature squared reconstruction error, residual-normalised."""
        residual = self.residuals(features)
        if self.residual_scale is not None:
            residual = residual / self.residual_scale
        return residual**2

    def score(self, features: np.ndarray) -> np.ndarray:
        """Mean squared reconstruction error per row -- the anomaly score."""
        return self.errors(features).mean(axis=1)

    def score_one(self, features: np.ndarray) -> float:
        return float(self.score(features)[0])

    def health_index(self, score: float) -> float:
        """Score mapped onto a 0-100 health scale, 50 exactly at the threshold.

        The exponent bends the curve so a nominal machine -- whose score sits
        around a third of the threshold -- reads in the high eighties rather
        than the low seventies, while a score several times over the threshold
        still collapses to zero. A dashboard that shows 70/100 for a perfectly
        healthy motor trains operators to ignore it.
        """
        if self.threshold <= 0.0:
            return 0.0
        ratio = max(score, 1e-12) / self.threshold
        return float(np.clip(100.0 * 2.0 ** (-(ratio**1.6)), 0.0, 100.0))

    def top_contributors(self, features: np.ndarray, k: int = 3) -> list[tuple[str, float]]:
        """Which features the model failed hardest to reconstruct, and by how much.

        A one-class model can only say "this is not normal". Attributing the
        error back to named features is what turns that into something an
        operator can act on.
        """
        per_feature = self.errors(features)[0]
        total = float(per_feature.sum()) or 1.0
        order = np.argsort(per_feature)[::-1][:k]
        return [(self.feature_names[i], float(per_feature[i] / total)) for i in order]

    # ---------------------------------------------------------------- #
    # Serialisation
    # ---------------------------------------------------------------- #

    def to_json(self) -> dict[str, Any]:
        return {
            "schema_version": SCHEMA_VERSION,
            "feature_names": list(self.feature_names),
            "scaler": {"mean": self.mean.tolist(), "scale": self.scale.tolist()},
            "layers": [layer.to_json() for layer in self.layers],
            "residual_scale": (
                None if self.residual_scale is None else self.residual_scale.tolist()
            ),
            "threshold": self.threshold,
            "metadata": self.metadata,
        }

    def save(self, path: str | Path) -> Path:
        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(self.to_json(), indent=1), encoding="utf-8")
        return path

    @classmethod
    def from_json(cls, payload: dict[str, Any]) -> AutoencoderBundle:
        version = payload.get("schema_version")
        if version != SCHEMA_VERSION:
            raise ValueError(
                f"model bundle schema {version} is not supported "
                f"(this build reads schema {SCHEMA_VERSION})"
            )
        scaler = payload["scaler"]
        residual_scale = payload.get("residual_scale")
        return cls(
            feature_names=tuple(payload["feature_names"]),
            mean=np.asarray(scaler["mean"], dtype=np.float64),
            scale=np.asarray(scaler["scale"], dtype=np.float64),
            layers=tuple(Layer.from_json(item) for item in payload["layers"]),
            threshold=float(payload["threshold"]),
            metadata=dict(payload.get("metadata", {})),
            residual_scale=(
                None if residual_scale is None else np.asarray(residual_scale, dtype=np.float64)
            ),
        )

    @classmethod
    def load(cls, path: str | Path) -> AutoencoderBundle:
        return cls.from_json(json.loads(Path(path).read_text(encoding="utf-8")))
